src/java/htsjdk/samtools/util/BlockCompressedInputStream.java

```diff
@@ -507,6 +541,119 @@ private static boolean preambleEqual(final byte[] preamble, final byte[] buf, fi
         }
         return true;
     }
+    
+    private static class DecompressedBlock {
+    	/**
+    	 * Decompressed block
+    	 */
+    	private final byte[] mBlock;
+    	/**
+    	 * Compressed size of block (the uncompressed size can be found using mBlock.length)
+    	 */
+    	private final int mBlockCompressedSize;
+    	/**
+    	 * Stream offset of start of block
+    	 */
+    	private final long mBlockAddress;
+    	/**
+    	 * Exception thrown (if any) when attempting to decompress block
+    	 */
+    	private final Exception mException;
+    	public DecompressedBlock(long blockAddress, byte[] block, int compressedSize) {
+    		mBlock = block;
+    		mBlockAddress = blockAddress;
+    		mBlockCompressedSize = compressedSize;
+    		mException = null;
+    	}
+    	public DecompressedBlock(long blockAddress, int compressedSize, Exception exception) {
+    		mBlock = new byte[0];
+    		mBlockAddress = blockAddress;
+    		mBlockCompressedSize = compressedSize;
+    		mException = exception;
+    	}
+    }
+    
+    private void startDecompressionThread() {
+    	if (mDecompressionThread == null) {
+    		mDecompressionRunnable = new DecompressionRunnable();
+    		mDecompressionThread = new Thread(mDecompressionRunnable, "DecompressBGZF");
+    		mDecompressionThread.setDaemon(true);
+    		mDecompressionThread.start();
+    	}
+    	if (mDecompressionRunnable.mResult == null) {
+    		mDecompressionRunnable.mResult = new ArrayBlockingQueue<DecompressedBlock>(1 + DecompressionRunnable.READ_AHEAD_DEPTH);
+    		assert(mDecompressionRunnable.mOperations.isEmpty());
+    		mDecompressionRunnable.mOperations.add(this);
+    		for (int i = 0; i < DecompressionRunnable.READ_AHEAD_DEPTH; i++) {
+    			mDecompressionRunnable.mOperations.add(this);
+    		}
+    	}
+    }
+    
+    private void terminateDecompressionThread() {
+    	if (mDecompressionThread != null) {
+	    	try {
+    			mDecompressionRunnable.mTerminate.set(true);
+    			mDecompressionRunnable.mOperations.clear();
+    			mDecompressionThread.interrupt();
+    			try {
+					mDecompressionThread.join();
+				} catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+				}	    		
+	    	} finally {
+	    		mDecompressionThread = null;
+	    		mDecompressionRunnable = null;
+	    	}
+    	}
+    }
+    
+    private void flushDecompressionThread() {
+    	if (mDecompressionThread != null) {
+    		mDecompressionRunnable.mResult = null;
+			mDecompressionRunnable.mOperations.clear();
+			mDecompressionThread.interrupt();
+    	}
+    }
+    
+	/**
+	 * Performs block decompression
+	 */
+	private class DecompressionRunnable implements Runnable {
+		/**
+		 * Number of block to eagerly decompress
+		 */
+		public static final int READ_AHEAD_DEPTH = 1;
```

Comment:
I've updated this to be driven from Defaults.NON_ZERO_BUFFER_SIZE as it already drives a range of buffers from this single value.